util/dvsim/Testplan.py

```diff
@@ -481,73 +482,35 @@ def get_stage_regressions(self):
             "tests": list(regressions[ms])
         } for ms in regressions]
 
-    def get_testplan_table(self, fmt="pipe"):
-        """Generate testplan table from hjson testplan.
+    def write_testplan_doc(self, output: TextIO) -> None:
+        """Write testplan documentation in markdown from the hjson testplan."""
 
-        fmt is either 'pipe' (markdown) or 'html'. 'pipe' is the name used by
-        tabulate to generate a markdown formatted table.
-        """
-        assert fmt in ["pipe", "html"]
-
-        # Map between the requested format and a pair (tabfmt, formatter) where
-        # tabfmt is the "tablefmt" argument for tabulate.tabulate and formatter
-        # converts the input Markdown text to something we can pass to the
-        # formatter.
-        fmt_configs = {
-            # For Markdown output, we pass the input text straight through
-            'pipe': ('pipe', lambda x: x),
-            # For HTML output, we convert the Markdown to HTML using the
-            # mistletoe library. The tablefmt argument should be 'unsafehtml'
-            # in this case because this already escapes things like '<' and
-            # don't want to double-escape them when tabulating.
-            'html': ('unsafehtml', mistletoe.markdown)
-        }
-        tabfmt, formatter = fmt_configs[fmt]
-
-        if self.testpoints:
-            lines = [formatter("\n### Testpoints\n")]
-            header = ["Stage", "Name", "Tests", "Description"]
-            colalign = ("center", "center", "left", "left")
-            table = []
-            for tp in self.testpoints:
-                desc = formatter(tp.desc.strip())
-
-                # tests is a list of strings. We want to insert them into a
-                # table and (conveniently) we can put one on each line in both
-                # Markdown and HTML mode by interspersing with '<br>' tags.
-                tests = "<br>\n".join(tp.tests)
-
-                table.append([tp.stage, tp.name, tests, desc])
-            lines += [
-                tabulate(table,
-                         headers=header,
-                         tablefmt=tabfmt,
-                         colalign=colalign)
-            ]
+        stages = {}
+        for tp in self.testpoints:
+            try:
```

Comment:
I would find it more clearer if `if t.stage not in stage` but that's just a personal preference.

Reply:
I think that something like the following is the approach that I've usually seen:
```python3
        stages = {}
        for tp in self.testpoints:
            stages.setdefault(tp.stage, list()).append(tp)
```

Reply:
ooooo `setdefault` is very nice

Reply:
Didn't know `setdefault`, will definitely use it in the future! Thanks @rswarbrick